### PR TITLE
Jennyf/redirect uri embedded webview

### PR DESCRIPTION
**Pull request by jennyf19**
need to support `urn:ietf:wg:oauth:2.0:oob` as a redirect uri for embedded webview
**Tested w/iOS and Android devices**
> (system browser and embedded webview)

**Tested w/iOS & Android on Xamarin-v2 sample (embedded webview)**

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ public sealed partial class PublicClientApplication : ClientApplicationBase
     {
 
         /// <summary>
-        /// Constructor to create application instance. This constructor is only available for Dekstop and NetCore apps
+        /// Constructor to create application instance. This constructor is only available for Desktop and NetCore apps
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:36]: >Desktop [](start = 95, length = 7)

:smile:  #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 2]
```diff
@@ -260,10 +258,16 @@ private void VerifyAuthorizationResult()
                     MsalErrorMessage.NoPromptFailedErrorMessage);
             }
 
+            if (_authorizationResult.Status == AuthorizationStatus.UserCancel)
+            {
+                throw new MsalClientException(_authorizationResult.Error,
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:262]: >_authorizationResult.Error [](start = 46, length = 26)

shouldn't this be the code for UserCancel ? #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 3]
```diff
@@ -1087,5 +1087,87 @@ public void GetUserTest()
...
+            MsalMockHelpers.ConfigureMockWebUI(ui);
+
+            try
+            {
+                AuthenticationResult result = await app.AcquireTokenAsync(TestConstants.Scope).ConfigureAwait(false);
+            }
+            catch (MsalClientException exc)
+            {
+                Assert.IsNotNull(exc);
+                Assert.AreEqual("authentication_canceled", exc.ErrorCode);
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1120]: >ErrorCode [](start = 63, length = 9)

I would assume this one should be user_canceled ? #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 4]
```diff
@@ -1087,5 +1087,87 @@ public void GetUserTest()
...
+                Assert.AreEqual("User canceled authentication", exc.Message);
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: >exc.Message [](start = 64, length = 11)

imo, it's not good to have anything witch looks like a sub error code in the message. this should be either in the error code or as a specialized exception #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 3 above)
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1120]: >ErrorCode [](start = 63, length = 9)

aside: imo the error code should be enum's - do you recall why it was agreed this should be a string? #Resolved

Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1120]: no...no idea why it's like this. #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 5]
```diff
@@ -1087,5 +1087,87 @@ public void GetUserTest()
...
+                Assert.IsNotNull(exc);
+                Assert.AreEqual("access_denied", exc.ErrorCode);
+                return;
+            }
+            finally
+            {
+                Assert.IsTrue(HttpMessageHandlerFactory.IsMocksQueueEmpty, "All mocks should have been consumed");
+            }
+
+            Assert.Fail("Should not reach here. Exception was not thrown.");
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1169]: >Assert.Fail("Should not reach here. Exception was not thrown."); [](start = 12, length = 64)

I believe you can attribute the test case to expect an exception. Though this works as well and I honestly have no preference, not sure though which pattern has been used in the code already. (same above) #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 6]
```diff
@@ -156,7 +154,7 @@ private Uri CreateAuthorizationUri(bool addVerifier = false, bool addState = fal
                 _state = Guid.NewGuid().ToString();
                 requestParameters[OAuth2Parameter.State] = _state;
             }
-
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:159]: >     [](start = 8, length = 4)

very minor: remove one of the empty lines #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 7]
```diff
@@ -60,8 +60,6 @@ internal class InteractiveRequest : RequestBase
             UIBehavior UIBehavior, IWebUI webUI)
             : base(authenticationRequestParameters)
         {
-            PlatformPlugin.PlatformInformation.ValidateRedirectUri(authenticationRequestParameters.RedirectUri,
-                authenticationRequestParameters.RequestContext);
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:64]: was this supposed to only be the case for embedded webview? #Resolved

**Comment by henrik-me** [comment]
>         public PublicClientApplication(string clientId, string authority, TokenCache userTokenCache) : base(clientId,

why is this one here an not in the other PCA file? #Resolved

---
Refers to: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:41 in 6436791. [](commit_id = 643679177b68c3e05fe5d7c4b03ef661dacdeea8, deletion_comment = False)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 7 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:64]: This only works on embedded webview, not system browser #Resolved

**Review by henrik-me — APPROVED** [review]
:shipit:

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 8]
```diff
@@ -260,10 +258,16 @@ private void VerifyAuthorizationResult()
                     MsalErrorMessage.NoPromptFailedErrorMessage);
             }
 
+            if (_authorizationResult.Status == AuthorizationStatus.UserCancel)
+            {
+                throw new MsalClientException(_authorizationResult.Error,
+                _authorizationResult.ErrorDescription);
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:263]: >_authorizationResult.ErrorDescription [](start = 16, length = 37)

also thinking the message should be specialized like the error code.  #Resolved

Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:263]: sounds good.  #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 5 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1169]: I agree, but was keeping the same format as the other tests in this section. #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 6 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:159]: done #Resolved

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 2 above)
Comment (jennyf19) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:262]: not sure what you mean by "the code for UserCancel". #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 3 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1120]: I changed this. made the ErrorCode `user_canceled_authentication` and removed the message, which had the same thing. lmk if you were thinking of something else. #Resolved

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 4 above)
Comment (jennyf19) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: i removed this in the message and made this the error code. #Resolved

**Comment by jennyf19** [comment]
>         public PublicClientApplication(string clientId, string authority, TokenCache userTokenCache) : base(clientId,

not sure why this is in it's own file

---
In reply to: [401989993](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#issuecomment-401989993) [](ancestors = 401989993)

---
Refers to: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:41 in 6436791. [](commit_id = 643679177b68c3e05fe5d7c4b03ef661dacdeea8, deletion_comment = False)

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/AdalError.cs [hunk 9]
```diff
@@ -55,7 +55,7 @@ public static class AdalError
         /// <summary>
         /// Authentication canceled.
         /// </summary>
-        public const string AuthenticationCanceled = "authentication_canceled";
+        public const string AuthenticationCanceled = "user_canceled_authentication";
```
Comment (henrik-me) [inline adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/AdalError.cs]: Not sure we can change this ne as it's in the public API... I was thinking we needed to add one extra for the new case introduced? #Resolved

File: core/src/MsalClientException.cs [hunk 10]
```diff
@@ -65,7 +65,7 @@ public class MsalClientException : MsalException
         /// <summary>
         /// Authentication canceled.
         /// </summary>
-        public const string AuthenticationCanceledError = "authentication_canceled";
+        public const string AuthenticationCanceledError = "user_canceled_authentication";
```
Comment (henrik-me) [inline core/src/MsalClientException.cs]: >"user_canceled_authentication" [](start = 58, length = 30)

same here #Resolved

File: core/src/UI/AuthorizationResult.cs [hunk 11]
```diff
@@ -51,7 +51,6 @@ internal AuthorizationResult(AuthorizationStatus status, string returnedUriInput
             if (Status == AuthorizationStatus.UserCancel)
             {
                 Error = MsalClientException.AuthenticationCanceledError;
-                ErrorDescription = MsalErrorMessage.AuthenticationCanceled;
```
Comment (henrik-me) [inline core/src/UI/AuthorizationResult.cs]: >ErrorDescription [](start = 16, length = 16)

Are you sure you meant to change existing behavior? #Resolved

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 12]
```diff
@@ -362,6 +371,14 @@ internal IWebUI CreateWebAuthenticationDialog(UIParent parent, UIBehavior behavi
             var requestParams = CreateRequestParameters(authority, scopes, user, UserTokenCache);
             requestParams.ExtraQueryParameters = extraQueryParameters;
 
+#if iOS || ANDROID
+            if(!parent.CoreUIParent.UseEmbeddedWebview)
+            {
+                PlatformPlugin.PlatformInformation.ValidateRedirectUri(requestParams.RedirectUri,
+                    requestParams.RequestContext);
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: minor: does this have to be 2 lines? (same above) #Resolved

**Comment by henrik-me** [comment]
>         public PublicClientApplication(string clientId, string authority, TokenCache userTokenCache) : base(clientId,

Currently this file makes no sense to me... perhaps move the function to the other file and delate this one?

---
In reply to: [402313370](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#issuecomment-402313370) [](ancestors = 402313370,401989993)

---
Refers to: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:41 in 6436791. [](commit_id = 643679177b68c3e05fe5d7c4b03ef661dacdeea8, deletion_comment = False)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 2 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:262]: right now you get the same error as before. I believe it should have the specific error code user cancelled... however if the current code calls that auth cancelled like you had it before then that is fine. IMO we should be more specific. Thus there is a difference to me in auth_cancelled and user_cancelled_auth. If however the meaning is the same, then the originally proposed error code is great.

Aside: I would generally still prefer to have the exceptions created in a factory. ensuring that the same exception type is created the same way everywhere.

---
In reply to: [199968924](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#discussion_r199968924) [](ancestors = 199968924)

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 3 above)
Comment (henrik-me) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs:1120]: pls. see other comments... probably easier to discuss in person.

---
In reply to: [199969044](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#discussion_r199969044) [](ancestors = 199969044)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 7 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs:64]: shouldn't it then still be enabled for the place where it makes sense?

---
In reply to: [199665688](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#discussion_r199665688) [](ancestors = 199665688)

**Comment by henrik-me** [comment]
>     public sealed partial class PublicClientApplication : ClientApplicationBase

this file should have been placed under a subfolder for the platform. then the folder structure would make this clear. alternatively all extra files should all be in a platfom sub-folder (similar to what already exists)

---
Refers to: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:32 in 077f67c. [](commit_id = 077f67c9158ad4114a609f9a21c377ef0b43c703, deletion_comment = False)

File: msal/tests/dev apps/XForms/XForms.iOS/AppDelegate.cs [hunk 13]
```diff
@@ -52,6 +52,9 @@ public override bool FinishedLaunching(UIApplication app, NSDictionary options)
             global::Xamarin.Forms.Forms.Init();
             LoadApplication(new App());
 
+            // Default system browser
+            App.UIParent = new UIParent();
```
Comment (henrik-me) [inline msal/tests/dev apps/XForms/XForms.iOS/AppDelegate.cs:56]: this is absolutely not obvious... has the same been done for Android? Also, perhaps this should be wrapped into a helper class names elegantly making the above comment unnecessary

**Comment by henrik-me** [comment]
:shipit:

File: msal/tests/dev apps/XForms/XForms.iOS/AppDelegate.cs (hunk 13 above)
Comment (jennyf19) [reply msal/tests/dev apps/XForms/XForms.iOS/AppDelegate.cs:56]: Yes, Android has something similar, but you need to pass in the [activity](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/dev/msal/tests/dev%20apps/XForms/XForms.Android/MainActivity.cs#L50). I like your suggestion though.

**Comment by jennyf19** [comment]
>     public sealed partial class PublicClientApplication : ClientApplicationBase

I agree. I will do this in another PR.

---
In reply to: [402880095](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1107#issuecomment-402880095) [](ancestors = 402880095)

---
Refers to: msal/src/Microsoft.Identity.Client/Features/PublicClient/PublicClientApplication.cs:32 in 077f67c. [](commit_id = 077f67c9158ad4114a609f9a21c377ef0b43c703, deletion_comment = False)

**Pull request merged**